[DOCU-876] Correlation ID Plugin: edit Nginx param (#3656)
Kong/docs.konghq.com: app/_hub/kong-inc/correlation-id/index.md

## FAQ

### Can I see my correlation IDs in my Kong logs?

You can see your correlation ID in the Nginx access or error logs if you edit your Nginx logging parameters.

To edit your Nginx parameters, do the following:

1. Locate [{{site.base_gateway}}'s template files](/gateway/latest/reference/configuration/#custom-nginx-templates) and make a copy of `nginx_kong.lua`. 

> 2. Add a `log_format` section:
> 
>      ```
>      log_format kvformat '<snipped_for_brevity> Kong-Request-ID="$sent_http_Kong_Request_ID" ';
>      ```

Reviewer: ```suggestion
2. Add a `log_format` section on the root level of the config file which includes the
  `$sent_http_Kong_Request_ID` variable. 
   On the following example, we created a new log format called `customformat`.
   It's a copy of the default `combined` log format, but the last line adds 
   `$sent_http_Kong_Request_ID`, preceded by the string `Kong-Request-ID=`. 
   Marking the variable this way is optional, but will make testing the feature easier.
   You can further customize the `log_format` by adding or removing
   [variables](http://nginx.org/en/docs/http/ngx_http_log_module.html):

     ```
     log_format customformat '$remote_addr - $remote_user [$time_local] '
                    '"$request" $status $body_bytes_sent  '
                    '"$http_referer" "$http_user_agent" '
                    'Kong-Request-ID="$sent_http_Kong_Request_ID"';
     ```

> 2. Reference it using: `access_log $PROXY_ACCESS_LOG kvformat;`

Reviewer: ```suggestion
2. Use your custom log format for the proxy access log phase. This means locating this line:
     ```
     access_log ${{PROXY_ACCESS_LOG}};
     ```
     And then modify it by adding the `customformat` format that we just created:
     ```
     access_log ${{PROXY_ACCESS_LOG}} customformat;
     ```
     Note: the file contains several `access_log` entries. You should only modify the line which uses `${{PROXY_ACCESS_LOG}}`, not the others.

## FAQ

### Can I see my correlation IDs in my Kong logs?

You can see your correlation ID in the Nginx access or error logs if you edit your Nginx logging parameters.

To edit your Nginx parameters, do the following:

1. Locate [{{site.base_gateway}}'s template files](/gateway/latest/reference/configuration/#custom-nginx-templates) and make a copy of `nginx_kong.lua`.
1. Add a `log_format` section on the root level of the config file which includes the
  `$sent_http_Kong_Request_ID` variable.

>    In the following example, we created a new log format named `customformat`.

Reviewer 2: ```suggestion

   In the following example, we create a new log format named `customformat`.
```

   It's a copy of the default `combined` log format, but the last line adds
   `$sent_http_Kong_Request_ID`, preceded by the string `Kong-Request-ID=`.
   Marking the variable this way is optional, and will make testing the feature easier.
   Further customize the `log_format` by adding or removing
   [variables](http://nginx.org/en/docs/http/ngx_http_log_module.html):

   ```
   log_format customformat '$remote_addr - $remote_user [$time_local] '
                 '"$request" $status $body_bytes_sent  '
                 '"$http_referer" "$http_user_agent" '
                 'Kong-Request-ID="$sent_http_Kong_Request_ID"';
   ```

> 1. Use your custom log format for the proxy access log phase. This means locating the following line:

Reviewer 2: ```suggestion
1. Use your custom log format for the proxy access log phase. Locate the following line:
```

   
   ```
   access_log ${PROXY_ACCESS_LOG};
   ```

>      Modifying it by adding the `customformat` format that we just created:

Reviewer 2: ```suggestion
     Modify it by adding the `customformat` format that we just created:
```

   ```

>    access_log ${{PROXY_ACCESS_LOG}} customformat;

Reviewer 2: ```suggestion
   access_log ${% raw %}{{PROXY_ACCESS_LOG}}{% endraw %} customformat;
```
Need these tags, otherwise jekyll tries to parse it as a variable.

Further note: is it accurate to use two curly braces in this example, or it is a typo? @kikito

   ```

     Note that the file contains several `access_log` entries. Only modify the line
     that uses `${PROXY_ACCESS_LOG}`.

2. Reload Kong:

   ```
   kong reload
   ```
  
3. Tail the access log:

   ```
   tail /usr/local/kong/logs/access.log
   ```

   You should now see Correlation ID entries in the access log.

> Learn more in [Custom Nginx templates & embedding Kong](/latest/configuration/#custom-nginx-templates--embedding-kong).

Reviewer 2: ```suggestion
Learn more in [Custom Nginx templates & embedding Kong](/gateway/latest/reference/configuration/#custom-nginx-templates--embedding-kong).
```

> ### Can I see my correlation ids in my Kong logs?

Reviewer 2: ```suggestion
### Can I see my correlation IDs in my Kong logs?
```

You can see your correlation ID in the Nginx access or error logs if you edit your Nginx logging parameters.

To edit your Nginx parameters, do the following:

> 1. In `/usr/local/share/lua/5.1/kong/templates/nginx_kong.lua`, add to a `log_format` section:

Reviewer 2: ```suggestion
1. Locate [{{site.base_gateway}}'s template files](/gateway/latest/reference/configuration/#custom-nginx-templates) and make a copy of `nginx_kong.lua`. 
2. Add a `log_format` section:
```
We should provide better guidelines on how to find the templates, and how to work with them. But since we don't want to have too much duplicate info, I think linking off to the config doc is sufficient.

     ```

>      log_format kvformat '<snipped_for_brevity> > > > > Kong-Request-ID="$sent_http_Kong_Request_ID" ';

Reviewer 3: ```suggestion
     log_format kvformat '<snipped_for_brevity> Kong-Request-ID="$sent_http_Kong_Request_ID" ';
```

>      log_format kvformat '<snipped_for_brevity> > > > > Kong-Request-ID="$sent_http_Kong_Request_ID" ';

Reviewer 2: Is there a particular place in the file that you need to add this section?

Author: @kikito addresses this in his edits

     ```

> 2. Reference it using: `access_log $PROXY_ACCESS_LOG kvformat;`

Reviewer 2: Similar to my previous question, where do you reference it? How do I know where to add an `access_log` phase - does it matter?

> 3. Reload Kong and tail the access log. You should see the entries for the Correlation ID in there.

Reviewer 2: Break this step into two:

> 3. Reload Kong:
>
>    ```sh
>    kong reload
>    ```
> 4. Tail the access log:
>
>    ```sh
>    tail logs/access.log
>    ```
>    You should see the entries for the Correlation ID in the log.

Not sure if that's actually the right log directory, I grabbed it from the custom example in the config ref.

Learn more in [Custom Nginx templates & embedding Kong](/latest/configuration/#custom-nginx-templates--embedding-kong).

> Note that you can also use this plugin along with one of the Logging plugins, or store the id on your backend.

Reviewer 2: ```suggestion
You can also use this plugin along with one of the [logging plugins](/hub/#logging), or store the ID on your backend.
```
